### PR TITLE
List tasks as txt

### PR DESCRIPTION
**Pull request by alenkacz**
List tasks as txt

Summary:
Last part of TasksController rewrite using Akka.HTTP. Prints out tasks in plaintext.

JIRA issues: MARATHON-7308

### CODE REVIEW

File: src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala [hunk 1]
```diff
@@ -43,6 +44,24 @@ object Directives extends AuthDirectives with LeaderDirectives with AkkaDirectiv
     }
   }
 
+  def accepts(mediaType: MediaType): Directive0 = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala]: Could you be so nice and use `accept` here https://github.com/mesosphere/marathon/blob/master/src/main/scala/mesosphere/marathon/api/akkahttp/SystemController.scala#L37 as well? Thanks a ton.

File: src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala [hunk 2]
```diff
@@ -45,15 +52,32 @@ class TasksController(
     asLeader(electionService) {
       authenticated.apply { implicit identity =>
         get {
-          pathEndOrSingleSlash {
-            listTasks()
+          (accepts(MediaTypes.`text/plain`) & pathEndOrSingleSlash) {
```
Comment (meln1k) [inline src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala]: Looks good, but do you think we can do the same thing using marshallers? It should be possible to define one marshaller based on several others using `Marshaller.oneOf`, and akka will do the rest of content negotiation.

File: src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala [hunk 3]
```diff
@@ -45,15 +52,32 @@ class TasksController(
...
+            listTasksJson() // when no accept header present, json is the default choice
           }
         }
       }
     }
   }
 
-  private def listTasks()(implicit identity: Identity): Route = {
+  @SuppressWarnings(Array("all")) // async/await
+  private def listTasksTxt()(implicit identity: Identity): Route = {
```
Comment (jeschkies) [inline src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala]: How come this method differs so much from `listTasksJson`?

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala [hunk 4]
```diff
@@ -161,6 +162,63 @@ class TasksControllerTest extends UnitTest with ScalatestRouteTest with Inside w
...
+      val app = AppDefinition("/foo".toRootPath, portDefinitions = Seq(PortDefinition(0), PortDefinition(0)))
+      val instance = TestInstanceBuilder.newBuilder(app.id).addTaskRunning().getInstance()
+
+      val tasksByApp = InstanceTracker.InstancesBySpec.forInstances(instance)
+      instanceTracker.instancesBySpec returns Future.successful(tasksByApp)
+
+      val rootGroup = createRootGroup(apps = Map(app.id -> app))
+      groupManager.rootGroup() returns rootGroup
+
+      assert(app.servicePorts.size > instance.appTask.status.networkInfo.hostPorts.size)
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: Please use the `should` matcher.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala [hunk 5]
```diff
@@ -161,6 +162,63 @@ class TasksControllerTest extends UnitTest with ScalatestRouteTest with Inside w
...
+
+      When("Getting the txt tasks index")
+      Get(Uri./).addHeader(Accept(MediaTypes.`text/plain`)) ~> controller.route ~> check {
+        Then("The status should be 200")
+        status should be(StatusCodes.OK)
+
+        responseAs[String] should include (s"${app.id.safePath}\t${app.servicePorts.head}")
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: Could we be explicit here? I think that's easier to understand.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala [hunk 6]
```diff
@@ -161,6 +162,63 @@ class TasksControllerTest extends UnitTest with ScalatestRouteTest with Inside w
...
+        status should be(StatusCodes.OK)
+
+        responseAs[String] should include (s"${authorizedApp.id.safePath}")
+        responseAs[String] should not include (s"${notAuthorizedApp.id.safePath}")
+      }
+    }
+
+    {
+      val controller = Fixture(authenticated = false).controller
+      behave like unauthenticatedRoute(forRoute = controller.route, withRequest = Get(Uri./).addHeader(Accept(MediaTypes.`text/plain`)))
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: How about

```scala
behave like unauthenticatedRoute(forRoute = controller.route, withRequest = Get(Uri./).addHeader(Accept(MediaTypes.`application/json`)))
```

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala [hunk 7]
```diff
@@ -161,6 +162,63 @@ class TasksControllerTest extends UnitTest with ScalatestRouteTest with Inside w
...
+
+      val rootGroup = createRootGroup(apps = Map(authorizedApp.id -> authorizedApp, notAuthorizedApp.id -> notAuthorizedApp))
+      f.groupManager.rootGroup() returns rootGroup
+
+      When("Getting the txt tasks index")
+      Get(Uri./).addHeader(Accept(MediaTypes.`text/plain`)) ~> f.controller.route ~> check {
+        Then("The status should be 200")
+        status should be(StatusCodes.OK)
+
+        responseAs[String] should include (s"${authorizedApp.id.safePath}")
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: Please use explicit assertions here. That is easier to understand.

File: src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala (hunk 2 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala]: well I thought about it but the thing is that the data from which one or the other response is created from is not the same - for json endpoint there is much more data fetched so the underlying model it different. That's why I did it this way, not the marshaller way. Does it make sense... ?

File: src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala (hunk 1 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala]: yes, I was actually planning to do that but in a separate PR once this one is approved (if you liked this approach - as it looks now you do :))

File: src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala (hunk 3 above)
Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala]: I would say that much less data is actually serialized to the text output than to json ... so even the processing and preparation of the data is a little bit different...

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala (hunk 5 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: what do you mean by being explicit?

File: src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala (hunk 1 above)
Comment (meln1k) [reply src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala]: Well, in general we don't need to explicitly filter accept headers as long as we have a proper marshaller. Akka http can do it transparently

Comment (alenkacz) [reply src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala]: I can try to rewrite ping using marshallers. I think there it should work...

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala (hunk 6 above)
Comment (alenkacz) [reply src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: added

**Comment by alenkacz** [comment]
@meln1k @jeschkies should be ready for re-review. Thanks!

File: src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala (hunk 1 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/api/akkahttp/Directives.scala]: I'd stay explicit here. I know that Akka should be able to do so. However, our API unfortunately does not seem to follow those rules exactly.

File: src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala (hunk 3 above)
Comment (jeschkies) [reply src/main/scala/mesosphere/marathon/api/akkahttp/v2/TasksController.scala]: Ok, thanks for clearing this up.

File: src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala [hunk 8]
```diff
@@ -135,12 +137,16 @@ class TasksControllerTest extends UnitTest with ScalatestRouteTest with Inside w
       behave like unauthenticatedRoute(forRoute = controller.route, withRequest = Get(Uri./))
     }
 
+    {
+      val controller = Fixture(authenticated = false).controller
+      behave like unauthenticatedRoute(forRoute = controller.route, withRequest = Get(Uri./).addHeader(Accept(MediaTypes.`application/json`)), " using text/json")
```
Comment (jeschkies) [inline src/test/scala/mesosphere/marathon/api/akkahttp/v2/TasksControllerTest.scala]: Why not move it into the same block as 
```scala
behave like unauthenticatedRoute(forRoute = controller.route, withRequest = Get(Uri./).addHeader(Accept(MediaTypes.`text/plain`)))
```

**Pull request merged**